lib/FIR/afforestation.cc

```diff
@@ -1057,59 +1085,21 @@ const parser::Format *FindReadWriteFormat(
   return FindReadWriteSpecifier<parser::Format>(specifiers);
 }
 
-static Expression *AlwaysTrueExpression() {
+static Expression AlwaysTrueExpression() {
   auto result{common::SearchTypes(
       evaluate::TypeKindVisitor<evaluate::TypeCategory::Logical,
           evaluate::Constant, bool>{1, true})};
   CHECK(result.has_value());
-  // TODO: this really ought to be hashconsed, but alas trees
-  return new evaluate::GenericExprWrapper(std::move(*result));
-}
-static Expression *LessThanOrEqualToComparison(
-    const semantics::Symbol *symbol, Expression *second) {
-  return second;  // FIXME
-}
-static Expression *GreaterThanOrEqualToComparison(
-    const semantics::Symbol *symbol, Expression *second) {
-  return second;  // FIXME
+  return {std::move(*result)};
 }
-static Expression *BuildLoopLatchExpression(
-    const std::optional<parser::LoopControl> &loopControl) {
-  if (loopControl.has_value()) {
-    return std::visit(
-        common::visitors{
-            [](const parser::LoopBounds<parser::ScalarIntExpr> &loopBounds) {
-              auto *loopVariable{loopBounds.name.thing.thing.symbol};
-              auto *second{loopBounds.upper.thing.thing->typedExpr.get()};
-              if (loopBounds.step.has_value()) {
-                auto *step{loopBounds.step->thing.thing->typedExpr.get()};
-                SEMANTICS_CHECK(step, "DO step expression missing");
-                if (evaluate::IsConstantExpr(step->v)) {
-                  auto stepValue{evaluate::ToInt64(step->v)};
-                  SEMANTICS_CHECK(stepValue != 0, "step cannot be zero");
-                  if (stepValue > 0) {
-                    return LessThanOrEqualToComparison(loopVariable, second);
-                  }
-                  return GreaterThanOrEqualToComparison(loopVariable, second);
-                }
-                // expr ::= (step > 0 && v <= snd) || (step < 0 && v >= snd)
-                return AlwaysTrueExpression();  // FIXME
-              }
-              return LessThanOrEqualToComparison(loopVariable, second);
-            },
-            [](const parser::ScalarLogicalExpr &scalarLogicalExpr) {
-              auto &expression{scalarLogicalExpr.thing.thing};
-              SEMANTICS_CHECK(
-                  expression->typedExpr.get(), "DO WHILE condition missing");
-              return expression->typedExpr.get();
-            },
-            [](const parser::LoopControl::Concurrent &concurrent) {
-              return AlwaysTrueExpression();  // FIXME
-            },
-        },
-        loopControl->u);
-  }
-  return AlwaysTrueExpression();
+
+// create an integer constant as an expression
+static Expression CreateConstant(int value) {
+  auto result{common::SearchTypes(
```

Comment:
I think this code will create an expression of type INTEGER(kind=value) with a constant value of 1, which is probably not what you want to do.

Reply:
Right. No, that's not correct.